benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py

```diff
@@ -0,0 +1,96 @@
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: S. Siso, STFC Daresbury Lab
+
+''' PSyclone transformation script showing the introduction of OpenMP for GPU
+directives into Nemo code. '''
+
+from psyclone.psyGen import TransInfo
+from psyclone.psyir.nodes import Loop, Assignment, CodeBlock
+from psyclone.domain.nemo.transformations import NemoAllArrayRange2LoopTrans
+from psyclone.transformations import TransformationError
+
+
+def trans(psy):
+    ''' Add OpenMP Target and Loop directives to Nemo loops over levels
+    in the provided PSy-layer.
+
+    :param psy: the PSy object which this script will transform.
+    :type psy: :py:class:`psyclone.psyGen.PSy`
+    :returns: the transformed PSy object.
+    :rtype: :py:class:`psyclone.psyGen.PSy`
+
+    '''
+    omp_target_trans = TransInfo().get_trans_name('OMPTargetTrans')
+    omp_loop_trans = TransInfo().get_trans_name('OMPLoopTrans')
+    # Disabling worksharing will produce the 'loop' directive which is better
+    # suited to map the work into the GPU
+    omp_loop_trans.omp_worksharing = False
+
+    print("Invokes found:")
+    for invoke in psy.invokes.invoke_list:
+        print(invoke.name)
+
+        # Convert all array implicit loops to explicit loops
+        explicit_loops = NemoAllArrayRange2LoopTrans()
+        for assignment in invoke.schedule.walk(Assignment):
+            explicit_loops.apply(assignment)
+
+        # Add the OpenMP directives in each loop
+        for loop in invoke.schedule.walk(Loop):
+            # TODO PSyclone/#1815: OMPTargetTrans does not exclude CodeBlock
+            # nodes.
+            if loop.loop_type == "levels" and not loop.walk(CodeBlock):
+                try:
+                    omp_target_trans.apply(loop)
+
+                    omp_loop_trans.apply(loop)
+                except TransformationError:
+                    # This loop can not be transformed, proceed to next loop
+                    continue
+
```

Comment:
Since we are here, to make it fast I went away with the loop_type and rely on the dependency analysis trying to paralelise each loop starting from the outer loop. I this code all outer loops but `do jt = 1, itn_count, 1` should succeed. Can you try if replacing this bit with the following gives better performance:
```
  # TODO PSyclone/#1815: OMPTargetTrans does not exclude CodeBlock
  # nodes.
  if loop.walk(CodeBlock):
      continue

  if loop.ancestor(Directive):
      # Skip if an outer loop is already parallelised
      continue

  try:
      omp_loop_trans.apply(loop)
      # Only add the target directive if the OMPLoop was successfully applied.
      omp_target_trans.apply(loop.parent.parent)
  except TransformationError as err:
      # This loop can not be transformed, proceed to next loop
      print("Loop not parallelised because:", str(err))
      continue
```

Reply:
Adding:
```
hoist_trans = HoistTrans()
for statement in invoke.schedule.walk(Statement):
    try:
        hoist_trans.apply(statement)
    except TransformationError as err:
        pass
```
before the OpenMP insertion loop may also help. (unfortunately repeated scalar assingments with the same value are not hoisted which would help here, I will open a ticket for this)

Reply:
I tried the first part of this but hit #1787 when I tried to collapse the loops. Without the collapsing the performance is no better.